File: internal/agent/shells.go
package agent

import (
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/remote-terminal-cloud/agent/internal/protocol"
)

var supportedShellTypes = map[protocol.ShellType]struct{}{
	protocol.ShellSystemDefault: {},
	protocol.ShellCmd:           {},
	protocol.ShellPowerShell:    {},
	protocol.ShellPwsh:          {},
	protocol.ShellBash:          {},
	protocol.ShellZsh:           {},
	protocol.ShellSh:            {},
}

type shellLaunch struct {
	executable string
	args       []string
	shellType  protocol.ShellType
}

func isSupportedShellType(value string) bool {
	_, ok := supportedShellTypes[protocol.ShellType(value)]
	return ok
}

func normalizeShellList(values []string) []protocol.ShellType {
	if len(values) == 0 {
		return nil
	}

	seen := make(map[protocol.ShellType]struct{}, len(values))
	result := make([]protocol.ShellType, 0, len(values))
	for _, value := range values {
		trimmed := strings.TrimSpace(value)
		if !isSupportedShellType(trimmed) {
			continue
		}

		shellType := protocol.ShellType(trimmed)
		if _, exists := seen[shellType]; exists {
			continue
		}

		seen[shellType] = struct{}{}
		result = append(result, shellType)
	}

	return result
}

func commandExists(command string) bool {
	if runtime.GOOS == "windows" {
		return runCommand("where.exe", []string{command}, 3*time.Second).OK
	}

	return runCommand("sh", []string{"-lc", "command -v " + command + " >/dev/null 2>&1"}, 3*time.Second).OK
}

func resolveExecutablePath(command string) string {
	trimmed := strings.TrimSpace(command)
	if trimmed == "" {
		return ""
	}

	if filepath.IsAbs(trimmed) {
		return trimmed
	}

	resolved, err := exec.LookPath(trimmed)
	if err == nil && strings.TrimSpace(resolved) != "" {
		return resolved
	}

	if runtime.GOOS == "windows" {
		result := runCommand("where.exe", []string{trimmed}, 3*time.Second)
		if result.OK {
			for _, line := range strings.Split(result.Stdout, "\n") {
				candidate := strings.TrimSpace(line)
				if candidate != "" {
					return candidate
				}
			}
		}
	}

	return trimmed
}

func detectAvailableShells() []protocol.ShellType {
	shells := []protocol.ShellType{protocol.ShellSystemDefault}

	if runtime.GOOS == "windows" {
		shells = append(shells, protocol.ShellCmd)
		if commandExists("powershell.exe") {
			shells = append(shells, protocol.ShellPowerShell)
		}
		if commandExists("pwsh.exe") {
			shells = append(shells, protocol.ShellPwsh)
		}
		return shells
	}

	candidates := []protocol.ShellType{
		protocol.ShellBash,
		protocol.ShellZsh,
		protocol.ShellSh,
		protocol.ShellPwsh,
	}
	for _, candidate := range candidates {
		if commandExists(string(candidate)) {
			shells = append(shells, candidate)
		}
	}

	return shells
}

func detectConfiguredAvailableShells(enabled []protocol.ShellType) []protocol.ShellType {
	detected := detectAvailableShells()
	if len(enabled) == 0 {
		return detected
	}

	enabledSet := make(map[protocol.ShellType]struct{}, len(enabled))
	for _, item := range enabled {
		enabledSet[item] = struct{}{}
	}

	filtered := make([]protocol.ShellType, 0, len(detected))
	for _, item := range detected {
		if _, ok := enabledSet[item]; ok {
			filtered = append(filtered, item)
		}
	}

	return filtered
}

func resolveEffectiveDefaultShell(configured protocol.ShellType, available []protocol.ShellType) protocol.ShellType {
	for _, item := range available {
		if item == configured {
			return configured
		}
	}
	for _, item := range available {
		if item == protocol.ShellSystemDefault {
			return protocol.ShellSystemDefault
		}
	}
	if len(available) > 0 {
		return available[0]
	}
	return configured
}

func resolveShellLaunch(requested protocol.ShellType, defaultShell protocol.ShellType) (shellLaunch, error) {
	normalized := requested
	if normalized == protocol.ShellSystemDefault {
		normalized = defaultShell
	}

	if runtime.GOOS == "windows" {
		switch normalized {
		case protocol.ShellSystemDefault, protocol.ShellCmd:
			comspec := os.Getenv("ComSpec")
			if strings.TrimSpace(comspec) == "" {
				comspec = "cmd.exe"
			}
			comspec = resolveExecutablePath(comspec)
			return shellLaunch{
				executable: comspec,
				args:       []string{"/d", "/k", "chcp 65001>nul"},
				shellType:  protocol.ShellCmd,
			}, nil
		case protocol.ShellPowerShell:
			return shellLaunch{
				executable: resolveExecutablePath("powershell.exe"),
				args:       []string{"-NoLogo", "-NoExit", "-Command", "[Console]::InputEncoding=[System.Text.UTF8Encoding]::new($false); [Console]::OutputEncoding=[System.Text.UTF8Encoding]::new($false); $OutputEncoding=[Console]::OutputEncoding; chcp 65001 > $null"},
				shellType:  protocol.ShellPowerShell,
			}, nil
		case protocol.ShellPwsh:
			return shellLaunch{
				executable: resolveExecutablePath("pwsh.exe"),
				args:       []string{"-NoLogo", "-NoExit", "-Command", "[Console]::InputEncoding=[System.Text.UTF8Encoding]::new($false); [Console]::OutputEncoding=[System.Text.UTF8Encoding]::new($false); $OutputEncoding=[Console]::OutputEncoding; chcp 65001 > $null"},
				shellType:  protocol.ShellPwsh,
			}, nil
		default:
			return shellLaunch{}, errors.New("shell is not supported on Windows")
		}
	}

	switch normalized {
	case protocol.ShellSystemDefault:
		systemShell := strings.TrimSpace(os.Getenv("SHELL"))
		if systemShell == "" {
			systemShell = "/bin/bash"
		}
		systemShell = resolveExecutablePath(systemShell)
		return shellLaunch{
			executable: systemShell,
			args:       []string{"-i"},
			shellType:  protocol.ShellSystemDefault,
		}, nil
	case protocol.ShellBash, protocol.ShellZsh, protocol.ShellSh:
		return shellLaunch{
			executable: resolveExecutablePath(string(normalized)),
			args:       []string{"-i"},
			shellType:  normalized,
		}, nil
	case protocol.ShellPwsh:
		return shellLaunch{
			executable: resolveExecutablePath("pwsh"),
			args:       []string{"-NoLogo"},
			shellType:  protocol.ShellPwsh,
		}, nil
	default:
		return shellLaunch{}, errors.New("shell is not supported on this platform")
	}
}

func cleanPath(input string) string {
	if strings.TrimSpace(input) == "" {
		return ""
	}
	return filepath.Clean(input)
}

File: src/shells.ts
import type { ShellType } from "@rtc/protocol";
import { runCommand } from "./process";

export const SUPPORTED_SHELL_TYPES = [
  "system-default",
  "cmd",
  "powershell",
  "pwsh",
  "bash",
  "zsh",
  "sh",
] as const satisfies readonly ShellType[];

export type SupportedShellType = (typeof SUPPORTED_SHELL_TYPES)[number];

const shellTypeSet = new Set<ShellType>(SUPPORTED_SHELL_TYPES);

export function isSupportedShellType(value: string | null | undefined): value is SupportedShellType {
  return value ? shellTypeSet.has(value as ShellType) : false;
}

async function commandExists(command: string): Promise<boolean> {
  if (process.platform === "win32") {
    const result = await runCommand("where.exe", [command]);
    return result.ok;
  }

  const result = await runCommand("sh", ["-lc", `command -v ${command} >/dev/null 2>&1`]);
  return result.ok;
}

export async function detectAvailableShells(): Promise<SupportedShellType[]> {
  const availableShells: SupportedShellType[] = ["system-default"];

  if (process.platform === "win32") {
    const candidates: SupportedShellType[] = ["cmd", "powershell", "pwsh"];
    for (const candidate of candidates) {
      if (candidate === "cmd") {
        availableShells.push(candidate);
        continue;
      }

      if (await commandExists(candidate === "powershell" ? "powershell.exe" : "pwsh.exe")) {
        availableShells.push(candidate);
      }
    }

    return availableShells;
  }

  const candidates: Array<{ command: string; shellType: SupportedShellType }> = [
    { command: "bash", shellType: "bash" },
    { command: "zsh", shellType: "zsh" },
    { command: "sh", shellType: "sh" },
    { command: "pwsh", shellType: "pwsh" },
  ];

  for (const candidate of candidates) {
    if (await commandExists(candidate.command)) {
      availableShells.push(candidate.shellType);
    }
  }

  return availableShells;
}

export async function detectConfiguredAvailableShells(
  enabledShellTypes: SupportedShellType[] | null,
): Promise<SupportedShellType[]> {
  const detectedShells = await detectAvailableShells();
  if (!enabledShellTypes) {
    return detectedShells;
  }

  const enabledShellSet = new Set(enabledShellTypes);
  return detectedShells.filter((shellType) => enabledShellSet.has(shellType));
}

export function resolveEffectiveDefaultShell(
  configuredDefaultShell: SupportedShellType,
  availableShells: SupportedShellType[],
): SupportedShellType {
  if (availableShells.includes(configuredDefaultShell)) {
    return configuredDefaultShell;
  }

  if (availableShells.includes("system-default")) {
    return "system-default";
  }

  return availableShells[0] ?? configuredDefaultShell;
}

export function resolveShellLaunch(
  requestedShellType: ShellType,
  defaultShellType: SupportedShellType,
): { executable: string; args: string[]; shellType: SupportedShellType; encoding?: string | null } {
  const normalizedShellType = requestedShellType === "system-default" ? defaultShellType : requestedShellType;

  const utf8Bootstrap = [
    "[Console]::InputEncoding = [System.Text.UTF8Encoding]::new($false)",
    "[Console]::OutputEncoding = [System.Text.UTF8Encoding]::new($false)",
    "$OutputEncoding = [Console]::OutputEncoding",
    "chcp 65001 > $null",
  ].join("; ");

  if (process.platform === "win32") {
    switch (normalizedShellType) {
      case "system-default":
      case "cmd":
        return {
          executable: process.env.ComSpec ?? "cmd.exe",
          args: ["/d", "/k", "chcp 65001>nul"],
          shellType: "cmd",
          encoding: "utf-8",
        };
      case "powershell":
        return {
          executable: "powershell.exe",
          args: ["-NoLogo", "-NoExit", "-Command", utf8Bootstrap],
          shellType: "powershell",
          encoding: "utf-8",
        };
      case "pwsh":
        return {
          executable: "pwsh.exe",
          args: ["-NoLogo", "-NoExit", "-Command", utf8Bootstrap],
          shellType: "pwsh",
          encoding: "utf-8",
        };
      default:
        throw new Error(`Shell ${normalizedShellType} is not supported on Windows.`);
    }
  }

  switch (normalizedShellType) {
    case "system-default":
      return {
        executable: process.env.SHELL ?? "/bin/bash",
        args: ["-i"],
        shellType: "system-default",
      };
    case "bash":
      return { executable: "bash", args: ["-i"], shellType: "bash" };
    case "zsh":
      return { executable: "zsh", args: ["-i"], shellType: "zsh" };
    case "sh":
      return { executable: "sh", args: ["-i"], shellType: "sh" };
    case "pwsh":
      return { executable: "pwsh", args: ["-NoLogo"], shellType: "pwsh" };
    default:
      throw new Error(`Shell ${normalizedShellType} is not supported on this platform.`);
  }
}